Move io_same_device hook to before attach_align_device hook on cpu_offload and disk_offload.

piEsposito opened the pull request:
Solves https://github.com/huggingface/accelerate/issues/767.

Moves the AlignDevicesHook with `io_same_device` from `accelerate.cpu_offload` and `accelerate.disk_offload` to before the `attach_align_device_hook`.

That way we can keep the changes on forward method for the whole module without deleting the hook we want to keep: the one with execution device and configurations on how to move the tensors between devices.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_768). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
The fix is not exactly right: by doing so, the hook that ensures the input and output of the model are on the same device is now erased. In your code sample in #767, since `x` is on the CPU, `net(x)` should also be on the CPU. This is not the case with your PR. The solution would be to write a util function that will:
- just add the hook if none is present
- extract the current hook if one is present and chain it with this hook using a `SequentialHook`.

This is slightly more advanced than the current PR, so let me know if you'd prefer for me to do it :-)

sgugger commented:
That works for me, though the name of the argument could simply be `append` :-)
Thanks for diving into this!

piEsposito commented:
@sgugger `append` it is.

sgugger commented on src/accelerate/hooks.py:
```diff
@@ -123,11 +123,21 @@ def add_hook_to_module(module: nn.Module, hook: ModelHook):
     Args:
         module (`torch.nn.Module`): The module to attach a hook to.
         hook (`ModelHook`): The hook to attach.
+        append (`bool`): Whether, if module already contains a hook, should chain the new one with SequentialHook
```
```suggestion
        append (`bool`, *optional*, defaults to `False`):
            Whether the hook should be chained with an existing one (if module already contains a hook) or not.
```

Also can't comment on line 119 but this should also be mentioned in the Note.

piEsposito replied:
Done.

sgugger commented on src/accelerate/hooks.py:
```diff
@@ -123,11 +123,21 @@ def add_hook_to_module(module: nn.Module, hook: ModelHook):
         be discarded).
     """
+
+    if append:
+        if hasattr(module, "_hf_hook") and (module._hf_hook is not None):
```
Those two ifs can be grouped.
Also the second test can be simplified in `getattr(module, "_hf_hook", None) is not None`

piEsposito replied:
Done, thanks for the great idea.

sgugger commented on src/accelerate/hooks.py:
```diff
@@ -123,11 +123,21 @@ def add_hook_to_module(module: nn.Module, hook: ModelHook):
+            old_hook = module._hf_hook
+            remove_hook_from_module(module)
+
+            hooks = [old_hook, hook]
+            hook = SequentialHook(*hooks)
```
```suggestion
            hook = SequentialHook(old_hook, hook)
```

piEsposito replied:
Done. Looks much better now.

sgugger reviewed (commented):
Very nice, thanks! Left a couple of nits, and I think you should still put the hook with io first: just tested locally and we still have the same issue of `net(x)` being on the wrong device since it runs second and the input was already moved.

piEsposito commented:
> Very nice, thanks! Left a couple of nits, and I think you should still put the hook with io first: just tested locally and we still have the same issue of `net(x)` being on the wrong device since it runs second and the input was already moved.

@sgugger I've just addressed your nits and moved the io hook to the top. Thanks for the review!

sgugger reviewed (approved):
Perfect, thanks! Re-tested locally and got the expected results for the code sample you shared in the issue.

piEsposito commented:
@sgugger, there is a test step that failed due to an http error when installing a lib. Can we try running it again or should I create a new commit to trigger the CI again?
